File: src/Data Structures/link list/ReorderList.cpp
#include <bits/stdc++.h>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode *reverseList(ListNode *head)
{
    ListNode *prev = NULL;
    ListNode *curr = head;
    ListNode *next;

    while (curr != NULL)
    {
        next = curr->next;
        curr->next = prev;
        prev = curr;
        curr = next;
    }

    return prev;
}

void reorderList(ListNode *head)
{
    ListNode *rightPointer = head;
    ListNode *leftPointer = head;

    while (rightPointer->next != NULL)
    {
        rightPointer = rightPointer->next->next != NULL ? rightPointer->next->next : rightPointer->next;
        leftPointer = leftPointer->next;
    }

    rightPointer = leftPointer->next;
    leftPointer->next = NULL;
    leftPointer = head;
    rightPointer = reverseList(rightPointer);

    ListNode *next1;
    ListNode *next2;

    while (leftPointer->next != NULL && rightPointer != NULL)
    {
        next1 = leftPointer->next;
        next2 = rightPointer->next;

        leftPointer->next = rightPointer;
        rightPointer->next = next1;

        rightPointer = next2;
        leftPointer = next1;
    }
}

int main()
{
    ListNode l1(3, NULL);
    ListNode l2(6, &l1);
    ListNode l3(9, &l2);

    ListNode *temp = &l3;
    int i = 0;

    temp = reverseList(&l3);

    while (temp != NULL && i != 5)
    {
        cout << temp->val << " -> ";
        temp = temp->next;
        i++;
    }
}

File: src/typescript/Data-Structures/Link-Lists/ReorderList.ts
// leetcode: 143. Reorder List https://leetcode.com/problems/reorder-list/ #Blind_75

class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
}

function reorderList(head: ListNode | null): void {
    let fastPointer = head, slowPointer = head;

    while (fastPointer.next) {
        fastPointer = fastPointer.next.next ? fastPointer.next.next : fastPointer.next;
        slowPointer = slowPointer.next;
    }

    fastPointer = slowPointer.next;
    fastPointer = reverseList(fastPointer);
    slowPointer.next = null;
    let temp = head;

    while (temp !== slowPointer && fastPointer) {
        let next1 = temp.next;
        let next2 = fastPointer.next;

        temp.next = fastPointer
        fastPointer.next = next1;

        fastPointer = next2;
        temp = next1;
    }

};

function reverseList(head: ListNode): ListNode {
    let prev = null, curr = head, next;

    while (curr) {
        next = curr.next;
        curr.next = prev;
        prev = curr
        curr = next;
    }
    return prev;
}